### PR TITLE
Add support for nesting_mode: "single"

### PR DESCRIPTION
**Pull request by thiagonache**
- Add support for `nesting_mode: "single"`
- Create two functions to emit datasources and emit resources since they are different now.
- Fix bugs in datasource generation.

### CODE REVIEW

**Comment by thiagonache** [comment]
@jpluscplusm this PR address `nesting_mode: "single"` while we don't agree on what to do about `nesting_mode: "list"` & `max_items: 1`

File: ratchet/ratchet.go [hunk 1]
```diff
@@ -15,23 +15,23 @@ func formatPrimitiveTypes(key, value string, typeAttributes gjson.Result) string
...
 	switch {
 	case typeAttributes.Get("required").Bool():
-		output = fmt.Sprintf("    %s!: %s", key, value)
+		output = fmt.Sprintf("%s!: %s", key, value)
 	case typeAttributes.Get("optional").Bool():
-		output = fmt.Sprintf("    %s?: %s", key, value)
+		output = fmt.Sprintf("%s?: %s", key, value)
 	default:
-		log.Fatalf("Attribute %q is neither required or optional", key)
+		log.Fatalf("Attribute %q is neither required or optional: %v", key, typeAttributes)
```
Comment (jpluscplusm) [inline ratchet/ratchet.go:22]: s/or/nor/

File: ratchet/ratchet.go [hunk 2]
```diff
@@ -89,11 +91,17 @@ func EmitAttribute(attrID string, entityID string, entityType string, terraformA
...
 }
 
-func EmitAttributes(entityID string, entityType string, terraformAttributes gjson.Result) string {
+func EmitDatasources(entityID string, terraformAttributes gjson.Result) string {
 	output := []string{}
 	required := map[string]gjson.Result{}
 	optional := map[string]gjson.Result{}
 	terraformAttributes.ForEach(func(attrID, terraformAttribute gjson.Result) bool {
+		if terraformAttribute.Get("optional").Bool() && terraformAttribute.Get("computed").Bool() {
+			//log.Fatalf("(datasource) %s: Attribute %q is both optional and computed: %v", entityID, attrID, terraformAttribute)
```
Comment (jpluscplusm) [inline ratchet/ratchet.go]: IIRC optional and computed is a legal combination. I /think/ it means "you can provide this, but it's available for runtime expressions even if you don't".

**Review by jpluscplusm — APPROVED** [review]

File: ratchet/ratchet.go (hunk 1 above)
Comment (thiagonache) [reply ratchet/ratchet.go:22]: thx

File: ratchet/ratchet.go (hunk 2 above)
Comment (thiagonache) [reply ratchet/ratchet.go]: makes sense!

**Pull request merged**